plutus-use-cases/test/Spec/Vesting.hs

```diff
@@ -1,30 +1,200 @@
+--   total of 60 lovelace over 20 blocks (20 lovelace can be taken out before
+--   that, at 10 blocks).
+vesting :: POSIXTime -> VestingParams
+vesting startTime =
+    VestingParams
+        { vestingTranche1 = VestingTranche (startTime + 10000) (Ada.lovelaceValueOf 20)
+        , vestingTranche2 = VestingTranche (startTime + 20000) (Ada.lovelaceValueOf 40)
+        , vestingOwner    = Ledger.pubKeyHash $ walletPubKey w1 }
+
+params :: VestingParams
+params = vesting (TimeSlot.scSlotZeroTime def)
+
+-- * QuickCheck model
+
+data VestingModel =
+  VestingModel { _vestedAmount :: Value
+               , _vested       :: [Wallet]
+               , _t1Slot       :: Slot
+               , _t2Slot       :: Slot
+               , _t1Amount     :: Value
+               , _t2Amount     :: Value
+               } deriving (Show, Eq)
+
+makeLenses 'VestingModel
+
+deriving instance Eq (ContractInstanceKey VestingModel w schema err)
+deriving instance Show (ContractInstanceKey VestingModel w schema err)
+
+instance ContractModel VestingModel where
+  data ContractInstanceKey VestingModel w schema err where
+    WalletKey :: Wallet -> ContractInstanceKey VestingModel () VestingSchema VestingError
+
+  data Action VestingModel = Vest Wallet
+                           | Retrieve Wallet Value
+                           | WaitUntil Slot
+                           deriving (Eq, Show)
+
+  initialState = VestingModel
+    { _vestedAmount = mempty
+    , _vested       = []
+    , _t1Slot       = TimeSlot.posixTimeToEnclosingSlot def $ vestingTrancheDate (vestingTranche1 params)
+    , _t2Slot       = TimeSlot.posixTimeToEnclosingSlot def $ vestingTrancheDate (vestingTranche2 params)
+    , _t1Amount     = vestingTrancheAmount (vestingTranche1 params)
+    , _t2Amount     = vestingTrancheAmount (vestingTranche2 params) }
+
+  perform handle _ cmd = case cmd of
+    Vest w -> do
+      callEndpoint @"vest funds" (handle $ WalletKey w) ()
+      delay 1
+
+    Retrieve w val -> do
+      callEndpoint @"retrieve funds" (handle $ WalletKey w) val
+      delay 2
+
+    WaitUntil slot -> void $ Trace.waitUntilSlot slot
+
+  nextState (Vest w) = do
+    let amount =  vestingTrancheAmount (vestingTranche1 params)
+               <> vestingTrancheAmount (vestingTranche2 params)
+    withdraw w amount
+    vestedAmount $~ (<> amount)
+    vested       $~ (w:)
+    wait 1
+
+  nextState (Retrieve w val) = do
+    slot   <- viewModelState currentSlot
+    t1     <- viewContractState t1Slot
+    t1v    <- viewContractState t1Amount
+    t2     <- viewContractState t2Slot
+    t2v    <- viewContractState t2Amount
+    amount <- viewContractState vestedAmount
+    let availableValue = mconcat $  [ t1v | slot > t1 ]
+                                 ++ [ t2v | slot > t2 ]
+        amountLeft = amount Numeric.- val
+        totalAmount = t1v <> t2v
+    when (  amountLeft `geq` (totalAmount Numeric.- availableValue)
```

Comment:
Could do with some comments, I'm struggling to understand what these conditions are stating.

Reply:
This entire module could do with some comments!

Reply:
C.f. recent commit but this condition states that the amount of value that is left in the contract after you retrieve is at least the amount that should be locked at that time (depending on where you are in relation to t1 or t2).